locked-asset/proxy_dex/src/energy_update.rs

```diff
@@ -8,25 +9,47 @@ static LOCKED_TOKEN_ID_STORAGE_KEY: &[u8] = b"lockedTokenId";
 
 #[elrond_wasm::module]
 pub trait EnergyUpdateModule: energy_query::EnergyQueryModule + utils::UtilsModule {
+    fn burn_locked_tokens_and_update_energy(
+        &self,
+        token_id: &TokenIdentifier,
+        token_nonce: Nonce,
+        token_amount: &BigUint,
+        user_address: &ManagedAddress,
+    ) {
+        if token_amount == &0u64 {
+            return;
+        }
+
+        self.deduct_energy_from_user(user_address, token_id, token_nonce, token_amount);
+        self.send()
+            .esdt_local_burn(token_id, token_nonce, token_amount);
+    }
+
     fn deduct_energy_from_user(
         &self,
         user: &ManagedAddress,
         token_id: &TokenIdentifier,
         token_nonce: u64,
         token_amount: &BigUint,
     ) {
+        let current_epoch = self.blockchain().get_block_epoch();
+        let mut energy = self.get_energy_entry(user);
+
         let energy_factory_addr = self.energy_factory_address().get();
-        let locked_token_id = self.get_locked_token_id(&energy_factory_addr);
-        if token_id != &locked_token_id {
-            return;
+        let new_locked_token_id = self.get_locked_token_id(&energy_factory_addr);
+        if token_id == &new_locked_token_id {
+            let attributes: LockedTokenAttributes<Self::Api> =
+                self.get_token_attributes(token_id, token_nonce);
+            energy.update_after_unlock_any(token_amount, attributes.unlock_epoch, current_epoch);
+        } else {
```

Comment:
For easier future updates, I would also put the check with legacy_locked_token_id here, otherwise it implies that it surely receives the legacy_locked_token, and I need to take a double check in the whole contract for that.